src/filters/ES6Ctors.ml

```diff
@@ -0,0 +1,225 @@
+				let this_before_super = has_this_before_super tf.tf_expr in
+				Option.may (fun e_this_access ->
+					activated := true;
+					let root_ref = ref cl_super in
+					Hashtbl.replace does_ctor_skipping cl.cl_path root_ref;
+					mark_needs_ctor_skipping cl_super root_ref e_this_access.epos;
+				) this_before_super
+			end else begin
+				(* if there was no ctor in the parent class, we still gotta call `super` *)
+				Hashtbl.add inject_super cl.cl_path cl; 
+			end
+		| _ -> ()
+	in
+	List.iter mark com.types;
+
+	if !activated then begin
+		(* just some helper common exprs *)
+		let e_false = (make_bool com.basic false null_pos) in 
+		let e_true = (make_bool com.basic true null_pos) in
+		let e_hx_ctor = (* this._hx_constructor *)
+			let ethis = mk (TConst TThis) t_dynamic null_pos  in
+			mk (TField (ethis, FDynamic ctor_method_name)) t_dynamic null_pos
+		in
+
+		let change t =
+			match t with
+			| TClassDecl ({ cl_constructor = Some ({ cf_expr = Some ({ eexpr = TFunction tf_ctor } as ctor_expr) } as cf_ctor) } as cl) ->
+				let does_ctor_skipping = try Some (Hashtbl.find does_ctor_skipping cl.cl_path) with Not_found -> None in
+
+				let add_hx_ctor_method () =
+					let cf_fun_ctor = mk_field ctor_method_name cf_ctor.cf_type cf_ctor.cf_pos null_pos in
+					cf_fun_ctor.cf_expr <- Some (replace_super_call ctor_expr);
+					cf_fun_ctor.cf_kind <- Method MethNormal;
+					cl.cl_ordered_fields <- cf_fun_ctor :: cl.cl_ordered_fields;
+					cl.cl_fields <- PMap.add cf_fun_ctor.cf_name cf_fun_ctor cl.cl_fields;
+				in
+
+				let make_hx_ctor_call e_skip_flag = (* this._hx_constructor(a,b,c) *)
+					let hxctor_call_args = List.map (fun (v,_) -> make_local v null_pos) tf_ctor.tf_args in
+					let hx_ctor_call = mk (TCall (e_hx_ctor, hxctor_call_args)) com.basic.tvoid null_pos in
+					if does_ctor_skipping <> None then
+						mk (TBlock [
+							mk (TBinop (OpAssign, e_skip_flag, e_true)) com.basic.tbool null_pos;
+							e_empty_super_call;
+							mk (TBinop (OpAssign, e_skip_flag, e_false)) com.basic.tbool null_pos;
+							hx_ctor_call
+						]) com.basic.tvoid null_pos
+					else
+						hx_ctor_call
+				in
+
+				let make_skip_flag root_cl = (* TopClass._hx_skip_constructor *)
+					let e_top = mk (TTypeExpr (TClassDecl root_cl)) t_dynamic null_pos in
+					mk (TField (e_top, FDynamic skip_ctor_flag_name)) com.basic.tbool null_pos
+				in
+
+				(match (try Some (Hashtbl.find needs_ctor_skipping cl.cl_path) with Not_found -> None) with
+				| Some root ->
+					let root = !root in
+
+					add_hx_ctor_method ();
+					
+					if does_ctor_skipping = None && cl != root then
+						(* for intermediate classes that support skipping but don't do skipping themselves, we can just remove the constructor altogether,
+						because the skipping logic is implemented in the parent constructor, and the actual constructor body is moved into _hx_constructor *)
+						cf_ctor.cf_expr <- None
```

Comment:
What if this class is instantiated somewhere directly?
Looks like the `_hx_constructor` will not be called.

Reply:
In this case, the body of the dummy constructor (the one that checks for skip and calls `_hx_constructor`) will already be present in the parent class (because if someone `does_ctor_skipping`, the `needs_ctor_skipping ` will propagate to the root. So we can save some generated code this way, e.g.

```haxe
class A {
	function new() {
		trace("hi from A");
	}
}

class B extends A {
	function new() {
		super();
		trace("hi from B");
	}
}

class C extends B {
	var i = 5;
}
```

will generate:

```js
class A {
	constructor() {
		if(A._hx_skip_constructor) {
			return;
		}
		this._hx_constructor();
	}
	_hx_constructor() {
		console.log("src/Main.hx:3:","hi from A");
	}
}
class B extends A {
	_hx_constructor() {
		super._hx_constructor();
		console.log("src/Main.hx:10:","hi from B");
	}
}
class C extends B {
	constructor() {
		A._hx_skip_constructor = true;
		super();
		A._hx_skip_constructor = false;
		this._hx_constructor();
	}
	_hx_constructor() {
		this.i = 5;
		super._hx_constructor();
	}
}
```

Then doing `new B()` will execute the constructor inherited from `A` which already contains all the required skipping handling.

Reply:
Got it